crash-recovery/README.md

```diff
@@ -0,0 +1,78 @@
+
+Hence it is crucial that the system should be able to recover from such crashes and bring itself to a consistent state.
+
+
+## Specification
+The ABCI proxy app adheres to the following state machine to commit a new block,
+
+```
+BEGIN BLOCK
+    ↓
+DELIVER TRANSACTION
+    ↓
+END BLOCK
+    ↓
+COMMIT
+```
+
+Here `BEGIN BLOCK` is used to indicate a new incoming block and `DEVLIER TRANSACTION` is used to iteratively deliver each transaction in this new block. `END BLOCK` is called by Tendermint when the set of transactions being commit-ed in the new block have been delivered (NOTE: the block height of the new block is passed as argument to `end_block`). Lastly, `COMMIT` is called and it is expected that all the state mutations are done during `COMMIT`. So, as previously discussed the transactions are stored and the block is added in the database.
+
+Since `COMMIT` api is mission critical and any crash during `COMMIT` can cause the system to be inconsistent so it is necessary that there should be a pre-commit state stored to the disk which should store the metadata of the state mutations being performed during actual `COMMIT` i.e. pre-commit object should at least store the following data,
+
+```json
+{ "id": "pre_commit",
+  "height": 10,
+  "transaction_ids": ["txid1", "txid2", "txidn"],
+}
+```
+
+The above object should be stored in a separate collection `pre_commit` during `END BLOCK`. Note that the `pre_commit` collection should have only 1 object with `'id': 'pre_commit'` because next write to `pre_commit` collection implies that the previous block was successfully `commit`ed. 
+
+### Recovering from crash
+In case of a crash following inconsistent object(s) might exists the database,
+
+- Zombie assets
+- Zombie metadata
+- Zombie transactions, inconsistent UTXO
+
+Following from the previous section, the metadata for `COMMIT` is stored in `pre_commit` collection which stores the block height and a list of transaction ids which will be written during `COMMIT`. Using this information we perform the following setups, 
+
+- Check if there is mismatch between latest height from `blocks` collection and `pre_commit` collection.
+- If there is no height mismatch then proceed normally.
+- If there is a height mismatch then the height difference should be at most 1 and latest block height retrieved from `blocks` collection should be less than that from `pre_commit` collection.
+- The block object retrieved from `pre_commit` collection provides the list of transaction ids which were to be commited.
+- If all the transaction ids from the above list exist in the `transactions` collection then the crash occurred after all the transactions were written and before the block could be written. In this case the block with new block height can be written to `blocks` collection using metadata from `pre_commit` collection.
```

Comment:
Do you think we can simplify and remove this condition? At the end, if there is a mismatch between the height in `blocks` and the one in the `pre_commit`, we can just run the cleanup function to remove the transactions in `pre_commit`, and wait for Tendermint to replay the block that has not be committed.

It generates a bit more load, but it simplifies the code in my opinion.